src/context/logging.ts

```diff
@@ -163,3 +164,89 @@ export class VoidLogger implements Logger {
 	info(category: string, message: string, data?: { [key: string]: any }): void {}
 }
 /* eslint-enable */
+
+interface LogEntry {
+	message: string;
+	level: string;
+	error?: string;
+}
+
+export class WshimLogger {
+	private logs: LogEntry[] = [];
+	private serviceToken: string;
+	private sampleRate: number;
+	private fetcher?: Fetcher;
```

Comment:
instead of a fetcher, this can be `typeof fetch`, and you assign the function directly in the constructor.